src/machine.cpp

```diff
@@ -2482,6 +2496,39 @@ uarch_interpreter_break_reason machine::run_uarch(uint64_t uarch_cycle_end) {
     return uarch_interpret(a, uarch_cycle_end);
 }
 
+interpreter_break_reason machine::log_step(uint64_t mcycle_count, const std::string &filename) {
+    if (!update_merkle_tree()) {
+        throw std::runtime_error{"error updating Merkle tree"};
+    }
+    hash_type root_hash_before;
+    get_root_hash(root_hash_before);
+    record_step_state_access a(*this, filename);
+    uint64_t mcycle_end{};
+    if (__builtin_add_overflow(a.read_mcycle(), mcycle_count, &mcycle_end)) {
+        mcycle_end = UINT64_MAX;
+    }
+    auto break_reason = interpret(a, mcycle_end);
+    a.finish();
+    hash_type root_hash_after;
+    get_root_hash(root_hash_after);
+    verify_step(root_hash_before, filename, mcycle_count, root_hash_after);
+    return break_reason;
+}
+
+void machine::verify_step(const hash_type &root_hash_before, const std::string &filename, uint64_t mcycle_count,
+    const hash_type &root_hash_after) {
+    auto data_length = os_get_file_length(filename.c_str(), "step log file");
+    auto *data = os_map_file(filename.c_str(), data_length, false /* not shared */);
+    replay_step_state_access a(data, data_length, root_hash_before);
+    uint64_t mcycle_end{};
+    if (__builtin_add_overflow(a.read_mcycle(), mcycle_count, &mcycle_end)) {
+        mcycle_end = UINT64_MAX;
+    }
+    interpret(a, mcycle_end);
```

Comment:
Why `break_reason` is not being returned? I think it wouldn't hurt to return it, and could be useful in some cases.